### PR TITLE
HFQuantizer implementation for compressed-tensors library

### PR DESCRIPTION
**Pull request by bfineran**
This PR adds an `HFQuantizer` for the [compressed-tensors](https://github.com/neuralmagic/compressed-tensors) library. 

Supported quantization features include:
* FP8, INT4, INT8 (for Q/DQ arbitrary precision is allowed for INT)
* Activation quantization (static)
* Dynamic per-token activation quantization
* Supports quantization of arbitrary layer types

Compressed tensors supports running in a Q/DQ format with transformer models and running compressed within vLLM (running compressed with transformers on the roadmap).

This initial PR includes a `HFQuantizer` and `QuantizationConfig` inplementations as well as a simple test.  Documentation is being added.

To run with this branch, compressed-tensors needs to be currently installed from a development branch (can be released to pypi prior to landing this PR pending support from the transformers team) - `pip install https://github.com/neuralmagic/compressed-tensors.git@rename_config`.

Happy to provide any further information needed.

Sample model load:
```python
from transformers import AutoModelForCausalLM
compressed_tensors_model = AutoModelForCausalLM.from_pretrained("nm-testing/tinyllama-oneshot-w4a16-group128-v3")
```

Suggested Reviewers:
@SunMarc @younesbelkada

### CODE REVIEW

**Comment by SunMarc** [comment]
Hi @bfineran, thanks for contributing and sorry for the delay ! From the PR, I see that we are decompressing the model at after loading the quantized model in order to run it. What would it take to run compressed model on transformers ? I don't mind merging this first but I want to make sure that we enable users to quantized model using compressed-library in the end. Happy to discuss more on how to collaborate together over slack if you are up to !

**Comment by bfineran** [comment]
Hi @SunMarc right now running compressed is WIP - we've prioritized a very flexible Q/DQ environment to enable a wide range of quantization settings and will likely roll out running compressed scenario by scenario.  Will reach out over slack to discuss more about the project and will also update with additional documentation soon.

File: src/transformers/quantizers/quantizer_compressed_tensors.py
Comment (SunMarc) [inline src/transformers/quantizers/quantizer_compressed_tensors.py:64]: I have a few points I would like to highlight: 
- It would be great to load the quantized weights into the model first. This will help a lot later when we want to actually run the compressed models. Right now, when I load the model, I get a lot of warnings. 
1) The first warning shows that `weights_packed` and `weight_shape` were ignored in the state dict because the model doesn't have these attributes. This can be fixed by removing them from the unexpected keys but it would be great to use them to load the compressed model in `create_quantized_param` or making sure that the model have these field after using `apply_quantization_config` : 
```
Some weights of the model checkpoint at nm-testing/tinyllama-oneshot-w4a16-group128-v3 were not used when initializing LlamaForCausalLM: ['model.layers.0.mlp.down_proj.weight_packed', 'model.layers.0.mlp.down_proj.weight_shape',
```
2) The second warning complains about missing keys in the checkpoint: 

```
Some weights of LlamaForCausalLM were not initialized from the model checkpoint at nm-testing/tinyllama-oneshot-w4a16-group128-v3 and are newly initialized: ['model.layers.0.mlp.down_proj.weight', 'model.layers.0.mlp.down_proj.weight_zero_point'
```
This can be solved by fixing the missing keys using the `update_missing_keys` method. Also, note that here we are complaining that `weight_zero_point` even though the compressed model doesn't use zero-point at all.

- I think it will be better if we can decompress the model just based on the loaded quantized model only ! I don't think we want to go through the state dict again to decompress the model. 

These parts are a bit hard to deal with, so feel free to ask questions ! 

Interested in your thoughts about these two points @ArthurZucker

File: tests/quantization/compressed_tensor/test_compressed_tensors.py
Comment (SunMarc) [inline tests/quantization/compressed_tensor/test_compressed_tensors.py:16]: Let's add more tests to fully cover all compression config ! This will help a lot to detect breaking changes in the future.

**Review by SunMarc — COMMENTED** [review]
Thanks for your great work @bfineran ! The PR is mostly good on my side. 
This new quantizer relies heavily on your own functions, so you will need to be a bit careful to not introduce breaking changes in the future (e.g. `apply_quantization_config` ) with transformers. However, I guess it can makes sense since compressed-tensors deals with many compressed format + no need to commit to transformers. 
There is just a blocker on the loading part of the quantizer that needs to be fixed !

File: src/transformers/quantizers/quantizer_compressed_tensors.py
Comment (bfineran) [inline src/transformers/quantizers/quantizer_compressed_tensors.py:32]:
```suggestion
```

File: src/transformers/quantizers/auto.py
Comment (bfineran) [reply src/transformers/quantizers/auto.py]: @SunMarc made this change - but it uncovered that since `compressed-tensors` is already used elsewhere (ie vllm) with the name `compressed-tensors`, this would be a breaking change. Would it be possible to keep the `-`? Could leave an in-line comment if that helps

File: src/transformers/quantizers/quantizer_compressed_tensors.py
Comment (bfineran) [reply src/transformers/quantizers/quantizer_compressed_tensors.py:69]: Great points, thanks!  Will go through the missing/unexpected keys lists to try and get these resolved.

We may be able to do this by splitting up some of the logic in decompression to load based on the quantized model specifically however do not currently have this in place.
Note that in decompression we shouldn't be re-loading the entire state dict since we're going through safe-tensors and only calling/decompressing one layer at a time.  Let me know if this would work or if you have further questions.

File: src/transformers/quantizers/auto.py
Comment (SunMarc) [reply src/transformers/quantizers/auto.py]: Let's leave it like that then if this will be breaking !

File: src/transformers/quantizers/quantizer_compressed_tensors.py
Comment (mgoin) [inline src/transformers/quantizers/quantizer_compressed_tensors.py:59]: Is there an issue with bfloat16? We should try to allow this for llama models

File: src/transformers/utils/quantization_config.py
Comment (mgoin) [inline src/transformers/utils/quantization_config.py:1104]: It would be good if we had a mechanism to pass through some kwargs here. This is already out of date with the recently added kv_cache_scheme https://github.com/neuralmagic/compressed-tensors/blob/d75e1a6beded5dac187553816f2f8c59781ffc23/src/compressed_tensors/quantization/quant_config.py#L142

File: tests/quantization/compressed_tensor/test_compressed_tensors.py
Comment (mgoin) [inline tests/quantization/compressed_tensor/test_compressed_tensors.py:46]:
```suggestion
            "quantized model should load a non-trivial scale into the state dict",
```

File: tests/quantization/compressed_tensor/test_compressed_tensors.py
Comment (mgoin) [reply tests/quantization/compressed_tensor/test_compressed_tensors.py:16]: @bfineran we could add some from the vllm [weight loading test](https://github.com/vllm-project/vllm/blob/3ac50b47d0f718364d05b1bcb93743e15be6a37c/tests/weight_loading/models.txt#L7-L15):
```
compressed-tensors, nm-testing/tinyllama-oneshot-w8w8-test-static-shape-change, main
compressed-tensors, nm-testing/tinyllama-oneshot-w8-channel-a8-tensor, main
compressed-tensors, nm-testing/tinyllama-oneshot-w8a8-dynamic-token-v2, main
compressed-tensors, nm-testing/tinyllama-oneshot-w8a8-channel-dynamic-token-v2, main
compressed-tensors, nm-testing/tinyllama-oneshot-w4a16-group128-v2, main
compressed-tensors, nm-testing/tinyllama-oneshot-w8a16-per-channel, main
compressed-tensors, nm-testing/Meta-Llama-3-8B-FP8-compressed-tensors-test, main
compressed-tensors, nm-testing/Phi-3-mini-128k-instruct-FP8, main
compressed-tensors, neuralmagic/Phi-3-medium-128k-instruct-quantized.w4a16, main
```

Llama and Phi-3 should cover dealing with unmerged and merged qkv layers

File: src/transformers/quantizers/quantizer_compressed_tensors.py
Comment (Satrat) [reply src/transformers/quantizers/quantizer_compressed_tensors.py:61]: No issue with bfloat16, we just recommend float16 as a default since that is what vLLM expects for the scale/zp

File: src/transformers/modeling_utils.py
Comment (SunMarc) [inline src/transformers/modeling_utils.py:4038]: Can be reverted

File: src/transformers/quantizers/quantizer_compressed_tensors.py
Comment (SunMarc) [inline src/transformers/quantizers/quantizer_compressed_tensors.py:77]: Do we still need to update these since we switched to the compressed model ?

File: src/transformers/quantizers/quantizer_compressed_tensors.py
Comment (SunMarc) [inline src/transformers/quantizers/quantizer_compressed_tensors.py:83]: wow nice api

**Comment by SunMarc** [comment]
Can you add the following to CompressedTensorsConfig ? This way we can print the quantization config

```py
    def to_diff_dict(self) -> Dict[str, Any]:
        """
        Removes all attributes from config which correspond to the default config attributes for better readability and
        serializes to a Python dictionary.
        Returns:
            `Dict[str, Any]`: Dictionary of all the attributes that make up this configuration instance,
        """
        config_dict = self.to_dict()

        # get the default config dict
        default_config_dict = CompressedTensorsConfig().to_dict()

        serializable_config_dict = {}

        # only serialize values that differ from the default config
        for key, value in config_dict.items():
            if value != default_config_dict[key]:
                serializable_config_dict[key] = value

        return serializable_config_dict
```

File: tests/quantization/compressed_tensor/test_compressed_tensors.py
Comment (SunMarc) [inline tests/quantization/compressed_tensor/test_compressed_tensors.py:74]: Maybe we should check for the expected outputs ? I know that when we modify the kernels, the output might not be the same and we will need to change. However, this is better in case something breaks

File: tests/quantization/compressed_tensor/test_compressed_tensors.py
Comment (SunMarc) [inline tests/quantization/compressed_tensor/test_compressed_tensors.py:17]: I would suggest adding a model for each method that can be used with compressed tensors.

File: src/transformers/quantizers/quantizer_compressed_tensors.py
Comment (SunMarc) [inline src/transformers/quantizers/quantizer_compressed_tensors.py:32]: We can't quantize on the fly with this quantizer, so we need to set that
```suggestion
    requires_calibration = True
```

File: docs/source/en/quantization/compressed_tensors.md
Comment (SunMarc) [inline docs/source/en/quantization/compressed_tensors.md:45]: maybe add a note saying that we can only load models that are already quantized

File: docs/source/en/quantization/compressed_tensors.md
Comment (SunMarc) [inline docs/source/en/quantization/compressed_tensors.md:30]: Could you also update the quantization overview page ? https://huggingface.co/docs/transformers/main/en/quantization/overview
This is a great document for users to find the right quantization tool for their usage !

**Review by SunMarc — APPROVED** [review]
Thanks for your patience and for iterating @bfineran @Satrat ! This looks good ! I left a few minor comments. I would be nice on your side to add test to check if everything works as expected also. We delegate a lot of stuff on your side as we are using the `apply_quantization_config` function !

File: docs/source/en/quantization/compressed_tensors.md
Comment (Satrat) [reply docs/source/en/quantization/compressed_tensors.md:30]: done!

File: tests/quantization/compressed_tensor/test_compressed_tensors.py
Comment (Satrat) [reply tests/quantization/compressed_tensor/test_compressed_tensors.py:86]: done!

File: docs/source/en/quantization/compressed_tensors.md
Comment (Satrat) [reply docs/source/en/quantization/compressed_tensors.md]: done!

File: tests/quantization/compressed_tensor/test_compressed_tensors.py
Comment (Satrat) [reply tests/quantization/compressed_tensor/test_compressed_tensors.py]: Yup, I added additional test cases for w4a16 and w8a16

**Comment by SunMarc** [comment]
Gentle ping @ArthurZucker

**Comment by SunMarc** [comment]
Can you rebase on main  @Satrat ? The failing might are probably due to that

**Comment by jvlinsta** [comment]
Eagerly awaiting this! Great work @neuralmagic team ;)

**Comment by Satrat** [comment]
> Can you rebase on main @Satrat ? The failing might are probably due to that

@SunMarc done! Looks like the tests are passing now after rebasing

**Comment by hyaticua** [comment]
Very eagerly awaiting this merge. Thanks to everyone in involved!

**Review by ArthurZucker — APPROVED** [review]
LGTM, I think it would be cool to improve the doc a bit and mention:
- this is only for inference (for now? 👀 )
- you cannot save back the loaded model (right?) 
- potential use-case (by that I mean loading a pre-quantized checkpoint is faster than loading non quantized, than quantizing). 
- for the format supported, I am curious about which FP8 and int4 and int8 (https://github.com/huggingface/optimum-quanto has int4 and int8, https://github.com/pytorch/ao/blob/main/torchao/prototype/low_bit_optim/subclass_4bit.py also has int4, would be nice to understand which "backend" it uses (of course, it's compressed-tensors, but a small explanation of that would help! ) and link to these types: basically help users!
🤗

File: src/transformers/utils/quantization_config.py
Comment (dsikka) [reply src/transformers/utils/quantization_config.py:1063]: @ArthurZucker this includes the different compression formats, depending on how the model is quantized/saved on disk, including: 
1. dense
2. int-quantized
3. float-quantized
4. pack-quantized
6. marlin-24

**Comment by dsikka** [comment]
@ArthurZucker Thank you for your feedback! We've updated the `compressed_tensors.md` addressing the aforementioned points.

**Review by SunMarc — APPROVED** [review]
Thanks for iterating on the docs ! LGTM ! Since I see that you release a new version of compressed_tensors, I'll merge this PR ! After this PR is merged @Satrat, could you modify the quantization_config of neural magic models to use `compressed_tensors` as we talked before ?

**Pull request merged**
